File: 백준/Gold/9370. 미확인 도착지/미확인 도착지.py
import sys
import heapq

def dijkstra(start):
    # dp[i][j] : i노드에서 j노드까지 가는 최단거리
    dp = [INF] * (n + 1)
    heap = []

    dp[start] = 0
    heapq.heappush(heap,(0, start))  # 순서 중요 (힙큐이기 때문에)

    #힙에 원소가 없을 때 까지 반복.
    while heap:
        now_dist, now_node = heapq.heappop(heap)

        # 현재 테이블보다 가중치가 큰 튜플이면 무시
        if dp[now_node] < now_dist:
            continue

        for next_node, dist in graph[now_node]:
            next_dist = dist + now_dist
            if next_dist < dp[next_node]:  # 유망하면
                dp[next_node] = next_dist
                heapq.heappush(heap,(next_dist, next_node))
    return dp


INF = sys.maxsize

T = int(sys.stdin.readline())
for _ in range(T):

    n, m, t = map(int, sys.stdin.readline().split())
    # s = 출발지, g, h = 무조건 지나는 경로
    s, g, h = map(int, sys.stdin.readline().split())

    graph = [[] for _ in range(n + 1)]

    # 다익스트라 외의 그냥 경로의 가중치를 저장해주는 그래프
    temp_graph = [[0]*(n+1) for _ in range(n + 1)]

    # 최단거리 테이블
    dp = [INF] * (n + 1)
    heap = []

    for _ in range(m):
        a, b, d = map(int, sys.stdin.readline().split())
        graph[a].append((b, d))
        graph[b].append((a, d))
        temp_graph[a][b] = d
        temp_graph[b][a] = d

    goal = [0]*t
    for i in range(t):
        goal[i] = int(sys.stdin.readline())

    start_s = dijkstra(s)  # s에서 출발
    start_g = dijkstra(g)  # g에서 출발
    start_h = dijkstra(h)  # h에서 출발

    goal.sort()
    for i in goal:
        # s -> g -> h -> i (목적지)인경우
        if start_s[i] == start_s[g] + temp_graph[g][h] + start_h[i]:
            print(i, end=' ')
        # s -> h -> g -> i (목적지)인경우
        elif start_s[i] == start_s[h] + temp_graph[h][g] + start_g[i]:
            print(i, end=' ')
    print()

File: 백준/Gold/9370. 미확인 도착지/미확인 도착지.cc
#include <bits/stdc++.h>
using namespace std;

typedef long long ll;
#define INF 1e9

/*

*/

int n, m, t, s, g, h;
vector<vector<pair<int,int>>> graph;

vector<int> dijkstra(int start){
    vector<int> dist(n + 1, INF);  // INF로 선언과 동시에 초기화
    priority_queue<pair<int, int>> pq;  // 우선순위 큐 선언

    dist[start] = 0;
    pq.push({0, start});  // 우선순위 큐에 넣기 (힙큐이므로 순서 중요!!)

    while (!pq.empty()) {
        int cur_dist = -pq.top().first;
        int cur_node = pq.top().second;
        pq.pop();

        // 현재 테이블보다 가중치가 큰 튜플이면 무시
        if (cur_dist > dist[cur_node]) continue;

        for (auto& edge : graph[cur_node]) {
            int nxt_node = edge.first;
            int nxt_dist = edge.second + cur_dist;
            
            if (nxt_dist < dist[nxt_node]) {  // 유망하면
                dist[nxt_node] = nxt_dist;
                pq.push({-nxt_dist, nxt_node});
            }
        }
    }

    return dist;  // start 노드로부터 각 노드까지 최단거리를 담은 벡터 리턴
}

void solve(){
    cin >> n >> m >> t;

    // 간선 그래프 초기화
    graph.clear();
    graph.resize(n+1);

    cin >> s >> g >> h;
    for (int i=0; i<m; i++){
        int a, b, d;
        cin >> a >> b >> d;
        graph[a].push_back({b, d});
        graph[b].push_back({a, d});
    }

    // 목적지 후보
    int goal[t];
    for (int i=0; i<t; i++){
        cin >> goal[i];
    }
    sort(goal, goal+t);

    auto startS = dijkstra(s);  // s에서 출발
    auto startG = dijkstra(g);  // g에서 출발
    auto startH = dijkstra(h);  // h에서 출발

    for (int d : goal){
        // s -> g -> h -> d (목적지)인경우
        if (startS[d]==startS[g]+startG[h]+startH[d]){
            cout << d << ' ';
        }
        // s -> h -> g -> d
        else if (startS[d]==startS[h]+startH[g]+startG[d]){
            cout << d << ' ';
        }
    }
    cout << '\n';
}

int main(){
    ios_base::sync_with_stdio(false);
    cin.tie(0);
    
    int T;
    cin >> T;
    while(T--) solve();
    
    return 0;
}